editor/groups_editor.cpp

```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
-
-	if (!selected) {
+void GroupsEditor::_set_group_checked(const String &p_name, bool checked) {
+	TreeItem *ti = tree->get_item_with_text(p_name);
+	if (!ti) {
 		return;
 	}
 
-	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
-	undo_redo->create_action(TTR("Remove from Group"));
+	ti->set_checked(0, checked);
+}
 
-	while (selected) {
-		Node *node = scene_tree->get_edited_scene_root()->get_node(selected->get_metadata(0));
-		undo_redo->add_do_method(node, "remove_from_group", selected_group);
-		undo_redo->add_undo_method(node, "add_to_group", selected_group, true);
+bool GroupsEditor::_has_group(const String &p_name) {
+	return global_groups.has(p_name) || scene_groups.has(p_name);
+}
 
-		selected = nodes_to_add->get_next_selected(selected);
+void GroupsEditor::_modify_group(Object *p_item, int p_column, int p_id, MouseButton p_mouse_button) {
+	if (p_mouse_button != MouseButton::LEFT) {
+		return;
 	}
 
-	undo_redo->add_do_method(this, "_group_selected");
-	undo_redo->add_undo_method(this, "_group_selected");
-	undo_redo->add_do_method(this, "emit_signal", "group_edited");
-	undo_redo->add_undo_method(this, "emit_signal", "group_edited");
+	if (!node) {
+		return;
+	}
 
-	// To force redraw of scene tree.
-	undo_redo->add_do_method(SceneTreeDock::get_singleton()->get_tree_editor(), "update_tree");
-	undo_redo->add_undo_method(SceneTreeDock::get_singleton()->get_tree_editor(), "update_tree");
+	TreeItem *ti = Object::cast_to<TreeItem>(p_item);
+	if (!ti) {
+		return;
+	}
 
-	undo_redo->commit_action();
+	switch (p_id) {
+		case COPY_GROUP: {
+			DisplayServer::get_singleton()->clipboard_set(ti->get_text(p_column));
+		} break;
+	}
 }
 
-void GroupDialog::_remove_filter_changed(const String &p_filter) {
-	_group_selected();
+void GroupsEditor::_filter_changed(const String &p_new_text) {
+	_update_tree();
 }
 
-void GroupDialog::_add_filter_changed(const String &p_filter) {
-	_group_selected();
-}
+void GroupsEditor::_load_scene_groups(Node *p_node) {
+	List<Node::GroupInfo> groups;
+	p_node->get_groups(&groups);
 
-void GroupDialog::_add_group_pressed(const String &p_name) {
-	_add_group(add_group_text->get_text());
-	add_group_text->clear();
-}
+	for (const GroupInfo &gi : groups) {
+		if (!gi.persistent) {
+			continue;
+		}
+
+		if (global_groups.has(gi.name)) {
+			continue;
+		}
+
+		bool is_editable = can_edit(p_node, gi.name);
+		if (scene_groups.has(gi.name)) {
+			scene_groups[gi.name] = scene_groups[gi.name] && is_editable;
+		} else {
+			scene_groups[gi.name] = is_editable;
+		}
+	}
 
-void GroupDialog::_add_group(String p_name) {
-	if (!is_visible()) {
-		return; // No need to edit the dialog if it's not being used.
+	for (int i = 0; i < p_node->get_child_count(); i++) {
+		_load_scene_groups(p_node->get_child(i));
 	}
+}
 
-	String name = p_name.strip_edges();
-	if (name.is_empty() || groups->get_item_with_text(name)) {
+void GroupsEditor::_update_groups() {
```

Comment:
Unless there is a reason to update the groups while the editor is invisible, this function can be optimized:
```C++
if (!is_visible_in_tree()) {
	groups_dirty = true;
	return;
}
groups_dirty = false
```
And then in _notification:
```C++
case NOTIFICATION_VISIBILITY_CHANGED: {
	if (groups_dirty && is_visible_in_tree()) {
		_update_groups();
	}
```
This is important, because it affects the whole editor. There is no need to update group list if the user is not currently using it.